test/integration/hds_integration_test.cc

```diff
@@ -219,6 +219,56 @@ class HdsIntegrationTest : public Grpc::VersionedGrpcClientIntegrationParamTest,
     }
   }
 
+  // check response has correct format and health response.
+  bool checkClusterEndpointHealthResponse(envoy::config::core::v3::HealthStatus healthy,
+                                          Network::Address::InstanceConstSharedPtr address,
+                                          int cluster, int locality, int endpoint) {
+    // Ensure that this grpc message is a health response.
+    if (response_.has_endpoint_health_response()) {
+      auto& health_response = response_.endpoint_health_response();
+
+      // Ensure that this response has a cluster available at the index.
+      if (health_response.cluster_endpoints_health_size() > cluster) {
+        auto& cluster_response = health_response.cluster_endpoints_health(cluster);
+
+        // Ensure that this response has a locality available at the index.
+        if (cluster_response.locality_endpoints_health_size() > locality) {
+          auto& locality_response = cluster_response.locality_endpoints_health(locality);
+
+          // Ensure that this response has a endpoint available at the index.
+          if (locality_response.endpoints_health_size() > endpoint) {
+            auto& endpoint_response = locality_response.endpoints_health(endpoint);
+
+            // Check to see if this endpoint has specified health status.
+            return checkEndpointHealthResponse(endpoint_response, healthy, address);
+          }
+        }
+      }
+    }
+
+    // Some field is missing, return false.
+    return false;
+  }
+
+  // wait until our response has desired health status for desired endpoint.
+  void waitForClusterHealthResponse(envoy::config::core::v3::HealthStatus healthy,
+                                    Network::Address::InstanceConstSharedPtr address, int cluster,
+                                    int locality, int endpoint) {
+    // Get some response.
+    ASSERT_TRUE(hds_stream_->waitForGrpcMessage(*dispatcher_, response_));
```

Comment:
ASSERT in a helper function doesn't work as expected -- you probably should make this a bool function and propagate a `return false;` and do the ASSERT in the test method.

Reply:
Thanks for this information, I had not realized that ASSERT statements are to stay in the test body to work properly. I went ahead and propagated the helper ASSERTs into the test body and just return a bool out of them.

Reply:
You can also use AssertionResult/Success/Failure as returns, grep for examples.